---
title: | 
       | <font size="16">`r params$title`</font>.
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  bookdown::html_document2:
    toc: true
    toc_float: true
    toc_depth: 4
    code_folding: hide
params:
  title: "stVsBulkComparison"
  runDirectory: '../'
  transcriptsFile: results/transcriptome/star/Heterocapsa/comparisons/P18363_103.mRNA.transcripts.stAndBulkAbundance.tsv
  genesFile: results/transcriptome/star/Heterocapsa/comparisons/P18363_103.mRNA.genes.stAndBulkAbundance.tsv
  doTranscripts: "TRUE"
---

```{r setup, include=FALSE, message=FALSE}
library(knitr)
library(kableExtra)
library(ggplot2)
library(cowplot)

opts_knit$set(root.dir=normalizePath(params$runDirectory), eval.after='fig.cap') # needed to get working dir right
knitr::opts_chunk$set(
	eval = TRUE,
	echo = FALSE,
	error = FALSE,
	fig.height = 3,
	fig.width = 6,
	message = FALSE,
	warning = FALSE,
	out.width = "100%",
	eval.after='fig.cap'
)
inline_hook <- function (x) {
  if (is.numeric(x)) {
    # ifelse does a vectorized comparison
    # If integer, print without decimal; otherwise print two places
    res <- ifelse(x == round(x),
      sprintf("%d", x),
      sprintf("%.2f", x)
    )
    paste(res, collapse = ", ")
  }
  else{
    x
  }
}
knit_hooks$set(inline = inline_hook)

showcode=TRUE
myfontsize = 12
options(knitr.table.format='html')
doTranscripts = ifelse(params$doTranscripts == "TRUE", TRUE, FALSE)
if(doTranscripts){

message("####### doTranscript is TRUE ######")
}else{
message("####### doTranscript is FALSE ######")
}
```

# Material & Methods

- Raw read counts (abundance) data from the analyses of data from the stpipeline ("st data") and bulk RNaseq data ("bulk data") were used as input. The st data read counts for each gene were summed over spots
- Pearson correlation coefficient $r$ and Spearman's rank correlation ooefficient $\rho$  was performed on genes with positive-valued data (see note below) using the R-function `cor`
- I have tentatively added correlation of log2-transformed count data also. This idea came from remembering that count data is almost always log2-transformed in expression analysis. The reason often stated for log2-transformation is that raw counts are very sensitive to outliers. I have noted that the (Pearson) correlation analysis is also affected heavily by outliers (at least fro some analyses), so it might make sense to use log2-transformation also here. HOIWEVER! This needs to be thought through properly, so treat the log2-transformed analyses with a bit of suspiciion for the time being!
- Pearson correlation  test for a *linear* relationship between variables, while Spearman correlation tests for a *monotonic* relationship between variables, i.e., variables may co-vary, but the correlation does not need to be a linear function. Hence it makes sense that it is the Pearson correlation that is affected by the log2-transformation.
- The correlation analysis was visualized using scatterplots, bith for the raw and the log2-transformed counts.
- A "correlation" (actually regression) line was estimated using a linear modl, $bulk\sim st$, in the plots (this was based on raw or log2-transformed positive values only, respectively). This is for a rough visualization of the correlation only (note that the regression would likely change sligthly if the reverse linear model was used, i.e., $st \sim bulk$).

## Notes:

- For the st results, read counts are missing for some genes. Similarly, for the bulk results, there are no missing read counts for genes, but there are genes with a zero-valued read count. It is possible that this also indicates missing values or values below detection (There appear to be no zero-valued read counts in the st results). In both these case, a (large) negative dummy value is assigned as the read count for that gene. Negative values are not included in the correlation analysis, but are shown, for completeness, in some plots.
- Using CPM-transformed data only changed the plots by changing the scale of the x- and y-axes, so it is not used in this analysis.

# Results

`r if(!doTranscripts){"<!--"}`
## Transcripts
```{r transcripts, eval=doTranscripts}

transcripts = read.delim(params$transcriptsFile)
transcripts[transcripts$count_st<=0, "count_st"] = -max(transcripts$count_st)/5 #-10000
transcripts[transcripts$count_bulk<=0, "count_bulk"] = -max(transcripts$count_bulk)/5 #-10000

posTranscripts = transcripts[transcripts$count_st > 0 & transcripts$count_bulk > 0,]

pearsonTranscripts = cor.test(posTranscripts$count_st, posTranscripts$count_bulk, method = "pearson", alternative = "two.sided")
spearmanTranscripts = cor.test(posTranscripts$count_st, posTranscripts$count_bulk, method = "spearman", alternative = "two.sided")

# Run correlation on log2-transformed values as well.
posTranscripts$log_count_st =log2(posTranscripts$count_st)
posTranscripts$log_count_bulk =log2(posTranscripts$count_bulk)
log_pearsonTranscripts = cor.test(posTranscripts$log_count_st, posTranscripts$log_count_bulk, method = "pearson", alternative = "two.sided")
log_spearmanTranscripts = cor.test(posTranscripts$log_count_st, posTranscripts$log_count_bulk, method = "spearman", alternative = "two.sided")

```

The result of the correlation analysis on transcripts level counts is shown in Figure \@ref(fig:plotTranscripts). For unlogged data, the Pearson correlation coefficient is $r = `r if(doTranscripts){pearsonTranscripts$estimate}` (p=`r if(doTranscripts){sprintf("%.2E",pearsonTranscripts$p.value)}`)$, the Spearman correlation is $\rho = `r if(doTranscripts){spearmanTranscripts$estimate}` (p=`r if(doTranscripts){sprintf("%.2E", spearmanTranscripts$p.value)}`)$.

For 2-logged data, the Pearson correlation coefficient is $r = `r if(doTranscripts){log_pearsonTranscripts$estimate}` (p=`r if(doTranscripts){sprintf("%.2E",log_pearsonTranscripts$p.value)}`)$, the Spearman correlation is $\rho = `r if(doTranscripts){log_spearmanTranscripts$estimate}` (p=`r if(doTranscripts){sprintf("%.2E", log_spearmanTranscripts$p.value)}`)$.

```{r plotTranscripts, eval=doTranscripts, fig.cap = paste0("Correlation plots for transcript-level read count analysis with (A) raw counts and (B) log2-transformed counts; the dotted blue line is the linear model-fitted regression line $count\\_bulk \\sim count\\_st$. Genes with missing data for counts have received a dummy value of -1000 (only shown in panel A).")}

p1 = ggplot(transcripts, aes(count_st, count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(data=posTranscripts,
	aes(count_st, count_bulk), method = "lm", se = FALSE, linetype="11", size=0.5) +
  theme_cowplot(8)
  
p2 = ggplot(posTranscripts, aes(log_count_st, log_count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(method = "lm", se = FALSE, linetype="11", size=0.5)+
  theme_cowplot(8)

plot_grid(p1,p2, labels = c("A","B"), label_size=8)
```
`r if(!doTranscripts){"-->"}`
## Genes

```{r genes}
origGenes = read.delim(params$genesFile)
genes=origGenes
genes[genes$count_st<=0, "count_st"] = -max(genes$count_st)/5 #-10000
genes[genes$count_bulk<=0, "count_bulk"] = -max(genes$count_bulk)/5 #-10000
posGenes = genes[genes$count_st > 0 & genes$count_bulk > 0,]

pearsonGenes = cor.test(posGenes$count_st, posGenes$count_bulk, method = "pearson", alternative = "two.sided")
spearmanGenes = cor.test(posGenes$count_st, posGenes$count_bulk, method = "spearman", alternative = "two.sided")

posGenes$log_count_st =log2(posGenes$count_st)
posGenes$log_count_bulk =log2(posGenes$count_bulk)
log_pearsonGenes = cor.test(posGenes$log_count_st, posGenes$log_count_bulk, method = "pearson", alternative = "two.sided")
log_spearmanGenes = cor.test(posGenes$log_count_st, posGenes$log_count_bulk, method = "spearman", alternative = "two.sided")

```

The result of the correlation analysis on genes level counts is shown in Figure \@ref(fig:plotGenes). 
For the unlogged data, the Pearson correlation coefficient is $r = `r pearsonGenes$estimate`\ (p=`r sprintf("%.2E",pearsonGenes$p.value)`)$, the Spearman correlation is $\rho = `r spearmanGenes$estimate`\ (p=`r sprintf("%.2E", spearmanGenes$p.value)`)$.

For the 2-logged data, the Pearson correlation coefficient is $r = `r log_pearsonGenes$estimate` (p=`r sprintf("%.2E",log_pearsonGenes$p.value)`)$, the Spearman correlation is $\rho = `r log_spearmanGenes$estimate` (`r sprintf("%.2E", log_spearmanGenes$p.value)`)$.


```{r plotGenes, fig.cap = paste0("Correlation plots for gene-level read count analysis with (A) raw counts and (B) log2-transformed counts; the dotted blue line is the linear model-fitted regression line $count\\_bulk \\sim count\\_st$. Genes with missing data for counts have received a dummy value of -1000 (only shown in panel A).")}

p3 = ggplot(genes, aes(count_st, count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(data=posGenes, aes(count_st, count_bulk), method = "lm", se = FALSE, linetype="11", size=0.5) +
  theme_cowplot(8)

p4 = ggplot(posGenes, aes(log_count_st, log_count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(data=posGenes, method = "lm", se = FALSE, linetype="11", size=0.5) +
  theme_cowplot(8)

plot_grid(p3, p4, labels = c("A","B"), label_size=8)

```


```{r simulate, echo=FALSE}

simResults = NULL #data.frame(pearson=list(), spearman=,list(), logPearson=list(), logSpearman=list())

N=100

for(j in seq(1,N)){
  sim = data.frame(genes = genes$genes, count_bulk=genes$count_bulk)
  sim$P = cumsum(sim$count_bulk/sum(sim$count_bulk))
  sim$count_st = 0
  
  idx = runif(n=sum(origGenes$count_st))
  t=table(sapply(idx, function(p){which(sim$P>=p)[1]}))
  sim[as.integer(names(t)), "count_st"] = t
  
  # (for(i in seq(1, sum(genes$count_st))){
  #   p=runif(1)
  #   idx = which(sim$P>=p)[1]-1
  #   sim[idx, "count_st"] = sim[idx, "count_st"]+1
  # }
  #   
  sim = sim[sim$count_st > 0 & sim$count_bulk > 0,]
  
  pearsonsim = cor.test(sim$count_st, sim$count_bulk, method = "pearson", alternative = "two.sided")$estimate
  spearmansim = cor.test(sim$count_st, sim$count_bulk, method = "spearman", alternative = "two.sided")$estimate
  
  sim$log_count_st = log2(sim$count_st)
  sim$log_count_bulk = log2(sim$count_bulk)
  log_pearsonsim = cor.test(sim$log_count_st, sim$log_count_bulk, method = "pearson", alternative = "two.sided")$estimate
  log_spearmansim = cor.test(sim$log_count_st, sim$log_count_bulk, method = "spearman", alternative = "two.sided")$estimate
  
  simResults =rbind(simResults, data.frame(pearsonsim,
                                           spearmansim, 
                                           log_pearsonsim, 
                                           log_spearmansim))
}


```


# Simulation

To investigate if the low read count in the st analysis could affect for the low correlation values, a simulation study was performed as follows.

1. The read count distribution over genes of the bulk RNAseq result, $B$ was used to approximate the cumulative probability distribution of sampling a read from gene. That is, 
  1. each gene's, $g$,  the read count was divided by the total number of reads and used as the probability of sampling a read, $r$, from that gene, $P[r|g]$.
  2. the cumulative $P[c|g]$ over genes (in the order they occurred in the result, this is arbitrary and will not affect generality) is the used as a cumulative distribution function $cdf$.
2. Let $N$ be the total read count of the sr analysis
3. Sample $N$ gene reads from $P[r|g]$ and for each gene, count the number of reads that has been sampled from it, this will be the simulated gene count, S$.
4. Perform Pearson and Spearman correlations on log2-transformed counts between $S$ and $B$ and record the correlation coefficients.

This procedure was repeated $`r N`$ times to obtain an approximated distribution of the respective correlation coefficient under the NULL model that the two compared read count samples are drawn from the same underlying distribution (Fig. \@ref(fig:histsim).


```{r plotsim, include=FALSE, fig.cap = paste0("Correlation plots for gene-level read count analysis with (A) raw counts and (B) log2-transformed counts; the dotted blue line is the linear model-fitted regression line $count\\_bulk \\sim count\\_st$. Genes with missing data for counts have received a dummy value of -1000 (only shown in panel A).")}

p5 = ggplot(sim, aes(count_st, count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(data=sim, aes(count_st, count_bulk), method = "lm", se = FALSE, linetype="11", size=0.5) +
  theme_cowplot(8)

p6 = ggplot(sim, aes(log_count_st, log_count_bulk)) +
  geom_point(size=0.5, color="red") + 
  geom_smooth(data=sim, method = "lm", se = FALSE, linetype="11", size=0.5) +
  theme_cowplot(8)

plot_grid(p5, p6, labels = c("A","B"), label_size=8)

```
```{r histsim, fig.cap = paste0("Histograms of simulated correlations of log2-transformed  gene-level read count analysis for (A) Pearson and (B) Spearman correlation values. The blue line indicate the observed correlation coefficient in the experimental data.")}

p5 = ggplot(simResults, aes(x=log_pearsonsim)) +
  geom_histogram() + 
  geom_vline(xintercept = log_pearsonGenes$estimate, color="blue")
p6 = ggplot(simResults, aes(x=log_spearmansim)) +
  geom_histogram() + 
  geom_vline(xintercept = log_spearmanGenes$estimate, color="blue")

plot_grid(p5, p6, labels = c("A","B"), label_size=8)

```